### PR TITLE
To replace base dialect segment class, must subclass or provide same stuff

### LINKED ISSUE #2902: Add internal check: Segments in "child" dialects must inherit from the parent dialect
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### Description

@OTooleMichael suggests:
>One should not be able to add a segment class matching a base name if it's not a subclass. Ala the replace decorator should complain

If this proves to be too strict, we could attempt to determine if the segment has any additional fields or methods beyond just `match_grammar` and `parse_grammar`, because that's mainly what we're trying to preserve. In the PR linked below, a `get_identifier()` method on ANSI's `CTEDefinitionSegment` was being lost in the Postgres dialect.

Reference: PR #2899

### Use case

Warn at dialect load time if a dialect violates this policy. 

### Dialect

Postgres was the inspiration, but really all dialects.

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by barrywhart**
### Brief summary of the change made
Fixes #2902

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/dialects/dialect_exasol.py [hunk 1]
```diff
@@ -3759,7 +3759,7 @@ class FunctionScriptStatementSegment(BaseSegment):
 
 
 @exasol_dialect.segment(replace=True)
-class StatementSegment(BaseSegment):
+class StatementSegment(ansi_dialect.get_segment("StatementSegment")):  # type: ignore
```
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_exasol.py:3762]: Is it better to do this? Or just add the `get_table_references` function to this Segment instead? We normally do the latter.

Comment (barrywhart) [reply src/sqlfluff/dialects/dialect_exasol.py:3762]: If the code is the same, I'd say this is better. Code reuse is one of the main benefits of inheritance, and copy paste. Especially in this case, where we'd be copying a good bit of code (about 10 lines).

Comment (tunetheweb) [reply src/sqlfluff/dialects/dialect_exasol.py:3762]: Oh definitely wasn’t suggesting copying and pasting the whole function. It’s whether you inherit the class or point the function to the ansi one. We tend to do the latter. But maybe that’s not the best way and this is better.

Comment (barrywhart) [reply src/sqlfluff/dialects/dialect_exasol.py:3762]: Got it. I don't have a strong feeling. One reason I did it this way is that it always works, even if the base class had several functions -- inheriting gives you access to them all, while the other way might require you to copy individual things, one by one.

TBH, I think it'd be good if we had a cleaner way to do this. Currently, the code is ugly either way. We could probably support something like this pretty easily:
```
class StatementSegment(ansi_dialect.StatementSegment)
```

That might allow us to omit the `# type: ignore` comments too.

Comment (tunetheweb) [reply src/sqlfluff/dialects/dialect_exasol.py:3762]: Yeah was curious why we needed to ignore the type here?

Comment (barrywhart) [reply src/sqlfluff/dialects/dialect_exasol.py:3762]: The type checker sees the function call `get_segment()` being used to specify the base class for the class we're creating, and it thinks, "oh, this is tricky. That function could return anything, so I can't analyze this the usual way, where I assume the base class is a known, specific thing." In reality, `get_segment()` is predictable, too, but the type checker doesn't know about it. I'm thinking if we changed this so we were using `.`, the type checker would assume it's more predictable. (It's not really!)

**Review by tunetheweb — APPROVED** [review]

**Pull request merged**